### PR TITLE
event mapping and normalization filters out valueless data

### PR DESCRIPTION
**Pull request by ndushay**
## Why was this change made?

Improves cocina event mapping and MODS originInfo normalization by filtering out valueless data (empty date values, empty originInfo elements ...)

### This Branch

```
Status (n=25000):
  Success:   24601 (98.404%)
  Different: 220 (0.88%)
  To Cocina error:     1 (0.004%)
  To Fedora error:     0 (0.0%)
  Missing (no descMetadata):     178 (0.712%)
```

#### roundtrip failures only in this branch
- kf303dc5137 - **roundtrip eventType changes (from production to publication)** and no empty dateOther
- qx562pf7510 - **roundtrip eventType changes (from manufacture to publication)** and no empty dateOther
- xv158sd4671 - **roundtrip eventType changes (from distribution to publication)** and no empty dateOther

#### other significant roundtrip differences between the branches for "Different"
(insignificant:  ordering of elements in mods or cocina)
- by352mp6687, fg349td3577, fv195wh3632, gv022qp2544, jk090bg7477, kq586bq1216, yh127sy5622 -- empty dateCreated no longer mapped/created
- nc238hg8745, yt337yg7382 - **roundtrip eventType  production vs publication**, empty dateCreated no longer mapped/created
- jz402xk5530 - **roundtrip eventType changes (from manufacture to publication)** and no empty dateOther;  **main branch loses displayLabel**, but this branch does not.
- vy980tx4948 
    - **roundtrip eventType changes (from manufacture to publication)** and no empty dateOther
    - **has data errors (Related resource has type and otherType)** (prob same across both branches?)
    - **a bunch of other roundtrip errors** (prob same across both branches?)

^^ all available in `ndushay/dor-services-app/results-new` folder on sdr-deploy.  Not sure for how long ...

### Main Branch

```
Status (n=25000):
  Success:   24604 (98.416%)
  Different: 217 (0.868%)
  To Cocina error:     1 (0.004%)
  To Fedora error:     0 (0.0%)
  Missing (no descMetadata):     178 (0.712%)
```

### CODE REVIEW

File: app/services/cocina/from_fedora/descriptive/event.rb [hunk 1]
```diff
@@ -373,26 +376,33 @@ def build_event(type, node_set, language_script = nil)
 
           notifier.warn('originInfo/dateOther missing eventType') unless type
 
-          {
-            type: type,
-            date: dates.presence
+          result = {
+            type: type
           }.compact
+          result[:date] = dates.compact if dates.present? && !dates.compact.empty? # deals with [nil]
```
Comment (justinlittman) [inline app/services/cocina/from_fedora/descriptive/event.rb:382]: Why not just `if dates.compact.present?`?

File: app/services/cocina/from_fedora/descriptive/event.rb [hunk 2]
```diff
@@ -373,26 +376,33 @@ def build_event(type, node_set, language_script = nil)
...
+        # rubocop:disable Metrics/CyclomaticComplexity
+        # rubocop:disable Metrics/PerceivedComplexity
         def build_structured_date(type, node_set)
           return nil if node_set.blank?
 
           common_attribs = common_date_attributes(node_set)
           return common_attribs.merge(value: node_set.join('/')) if etdf_range?(node_set, common_attribs[:encoding])
 
           dates = node_set.map do |node|
+            next if node.text.blank? && node.attributes.size.zero?
```
Comment (justinlittman) [inline app/services/cocina/from_fedora/descriptive/event.rb:395]: `node.attributes.empty?`?

**Review by justinlittman — APPROVED** [review]
Approve pending consideration of comments.

File: app/services/cocina/from_fedora/descriptive/event.rb (hunk 1 above)
Comment (ndushay) [reply app/services/cocina/from_fedora/descriptive/event.rb:382]: that is a better way, yes.  :-)

File: app/services/cocina/from_fedora/descriptive/event.rb (hunk 2 above)
Comment (ndushay) [reply app/services/cocina/from_fedora/descriptive/event.rb:395]: this is also a better way!

**Comment by ndushay** [comment]
suggestions taken - thank you.

**Comment by ndushay** [comment]
I would like arcadia to sign off on the changes in the roundtripping ...

File: app/services/cocina/mods_normalizers/origin_info_normalizer.rb [hunk 3]
```diff
@@ -33,7 +36,22 @@ def normalize
...
+      # must be called before remove_empty_origin_info
+      def remove_empty_origin_info_dates
+        DATE_FIELDS.without('dateOther').each do |date_field|
+          ng_xml.root.xpath("//mods:originInfo/mods:#{date_field}", mods: ModsNormalizer::MODS_NS).each do |date_node|
+            date_node.remove if date_node.content.blank?
+          end
+        end
+
+        # we can also remove dateOther if it has no type attribute
+        ng_xml.root.xpath('//mods:originInfo/mods:dateOther[not(@type)]', mods: ModsNormalizer::MODS_NS).each do |date_node|
```
Comment (arcadiafalcone) [inline app/services/cocina/mods_normalizers/origin_info_normalizer.rb:48]: dateOther can be removed if it doesn't have content regardless of whether it has an attribute value.

File: spec/services/cocina/mapping/descriptive/mods/origin_info_date_spec.rb [hunk 4]
```diff
@@ -1870,4 +1870,134 @@
...
+                  type: 'organization',
+                  role: [
+                    {
+                      value: 'distributor',
+                      code: 'dst',
+                      uri: 'http://id.loc.gov/vocabulary/relators/dst',
+                      source: {
+                        code: 'marcrelator',
+                        uri: 'http://id.loc.gov/vocabulary/relators/'
+                      }
```
Comment (arcadiafalcone) [inline spec/services/cocina/mapping/descriptive/mods/origin_info_date_spec.rb:1923]: Always use publisher as the contributor role regardless of the event type.

File: spec/services/cocina/mods_normalizers/origin_info_normalizer_spec.rb [hunk 5]
```diff
@@ -674,6 +674,184 @@
...
+      let(:mods_ng_xml) do
+        Nokogiri::XML <<~XML
+          <mods #{MODS_ATTRIBUTES}>
+            <originInfo>
+              <dateIssued encoding="w3cdtf" keyDate="yes"/>
+            </originInfo>
+          </mods>
+        XML
+      end
+
```
Comment (arcadiafalcone) [inline spec/services/cocina/mods_normalizers/origin_info_normalizer_spec.rb:728]: Ignoring attributes with value if the element has no content applies to all of the date fields, not just dateIssued.

File: spec/services/cocina/mods_normalizers/origin_info_normalizer_spec.rb [hunk 6]
```diff
@@ -674,6 +674,184 @@
...
 
+  describe 'empty originInfo date elements' do
+    context 'when dateCreated' do
+      # based on jw174hd9042, vy673zb2925
+      let(:mods_ng_xml) do
+        Nokogiri::XML <<~XML
+          <mods #{MODS_ATTRIBUTES}>
+            <originInfo>
+              <dateCreated></dateCreated>
+            </originInfo>
```
Comment (arcadiafalcone) [inline spec/services/cocina/mods_normalizers/origin_info_normalizer_spec.rb:685]: The empty element may also be in the form `<dateCreated/>` or `<dateCreated />`.

File: app/services/cocina/mods_normalizers/origin_info_normalizer.rb (hunk 3 above)
Comment (ndushay) [reply app/services/cocina/mods_normalizers/origin_info_normalizer.rb:48]: i'll make a ticket for this.  I think we use the @type attribute to determine stuff ... now, at least.  Is that going away with the new mappings?

**Review by ndushay — COMMENTED** [review]
it may be that there needs to be more normalization to counteract the changes here, but in my attempts to provide that, I never got the roundtripping to match what we have currrently ... I would like Arcadia to specifically weigh in on those differences as described in the description part of the PR.

**Comment by arcadiafalcone** [comment]
kf303dc5137 - New rule added to #2128 addresses this. (I'll probably also refresh all the records produced by the MODS 3.4 stylesheet as they're not very consistent with later data.)

qx562pf7510, xv158sd4671  - Original eventType should be retained per the latest mapping.

Dropping the empty dateOther and dateCreated is fine.

nc238hg8745, yt337yg7382  - eventType should be production.

jz402xk5530  - New #2128 rule addresses this.

vy980tx4948 - Original eventType should be retained per the latest mapping.

**Comment by ndushay** [comment]
Add to new roundtrip tests for eventType:  

- qx562pf7510, xv158sd4671 
- nc238hg8745, yt337yg7382 - eventType should be production.
- vy980tx4948 - Original eventType should be retained per the latest mapping.

**Comment by ndushay** [comment]
Merged this as JLitt and Arcadia signed off

File: app/services/cocina/mods_normalizers/origin_info_normalizer.rb (hunk 3 above)
Comment (ndushay) [reply app/services/cocina/mods_normalizers/origin_info_normalizer.rb:48]: issue #2240

File: spec/services/cocina/mapping/descriptive/mods/origin_info_date_spec.rb (hunk 4 above)
Comment (ndushay) [reply spec/services/cocina/mapping/descriptive/mods/origin_info_date_spec.rb:1923]: issue #2240

File: spec/services/cocina/mods_normalizers/origin_info_normalizer_spec.rb (hunk 5 above)
Comment (ndushay) [reply spec/services/cocina/mods_normalizers/origin_info_normalizer_spec.rb:728]: issue #2240

File: spec/services/cocina/mods_normalizers/origin_info_normalizer_spec.rb (hunk 6 above)
Comment (ndushay) [reply spec/services/cocina/mods_normalizers/origin_info_normalizer_spec.rb:685]: issue #2240

**Pull request merged**